Improve RTCRtpSender::setParameters spec compliance

youennf opened the pull request:
#### b6965cf9b33991b4d175e1f2ed2389641014dfe4
<pre>
Improve RTCRtpSender::setParameters spec compliance
<a href="https://bugs.webkit.org/show_bug.cgi?id=242853">https://bugs.webkit.org/show_bug.cgi?id=242853</a>
rdar://problem/97183896

Reviewed by Alex Christensen.

Add parameter checks as per spec.
In particular, we only fill cname for sender and not receiver.
We also validate the parameters provided to setParameters.

Covered by rebased tests.

* LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCRtpParameters-codecs-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCRtpParameters-degradationPreference-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCRtpParameters-encodings-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCRtpParameters-headerExtensions-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCRtpParameters-rtcp-expected.txt:
* Source/WebCore/Modules/mediastream/RTCRtpSender.cpp:
(WebCore::RTCRtpSender::getParameters):
(WebCore::validateModifiedParameters):
(WebCore::RTCRtpSender::setParameters):
* Source/WebCore/Modules/mediastream/RTCRtpSender.h:
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCRtpSenderBackend.cpp:
(WebCore::validateModifiedParameters):
(WebCore::LibWebRTCRtpSenderBackend::setParameters):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCUtils.cpp:
(WebCore::toRTCRtpParameters):
(WebCore::toRTCRtpSendParameters):

Canonical link: <a href="https://commits.webkit.org/252787@main">https://commits.webkit.org/252787@main</a>
</pre>

achristensen07 commented on Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCRtpSenderBackend.cpp:
```diff
@@ -123,6 +123,55 @@ RTCRtpSendParameters LibWebRTCRtpSenderBackend::getParameters() const
     return toRTCRtpSendParameters(*m_currentParameters);
 }
 
+static bool validateModifiedParameters(const RTCRtpSendParameters& newParameters, const RTCRtpSendParameters& oldParameters)
```
It seems like this code would be better placed in RTCRtpSendParameters::operator==

youennf replied:
RTCRtpSendParameters::operator== would need to check all fields (read-only and read/write).
In our case, we are only comparing read-only properties.

youennf committed:
Improve RTCRtpSender::setParameters spec compliance
https://bugs.webkit.org/show_bug.cgi?id=242853
rdar://problem/97183896

Reviewed by Alex Christensen.

Add parameter checks as per spec.
In particular, we only fill cname for sender and not receiver.
We also validate the parameters provided to setParameters.

Covered by rebased tests.

* LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCRtpParameters-codecs-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCRtpParameters-degradationPreference-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCRtpParameters-encodings-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCRtpParameters-headerExtensions-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webrtc/RTCRtpParameters-rtcp-expected.txt:
* Source/WebCore/Modules/mediastream/RTCRtpSender.cpp:
(WebCore::RTCRtpSender::getParameters):
(WebCore::validateModifiedParameters):
(WebCore::RTCRtpSender::setParameters):
* Source/WebCore/Modules/mediastream/RTCRtpSender.h:
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCRtpSenderBackend.cpp:
(WebCore::validateModifiedParameters):
(WebCore::LibWebRTCRtpSenderBackend::setParameters):
* Source/WebCore/Modules/mediastream/libwebrtc/LibWebRTCUtils.cpp:
(WebCore::toRTCRtpParameters):
(WebCore::toRTCRtpSendParameters):

Canonical link: https://commits.webkit.org/252787@main
